src/pure/pure.ml

```diff
@@ -4,11 +4,121 @@ open Extra
+    | Syntax.P_Wrap term -> qidents_of_p_term term
+    | Syntax.P_Expl term -> qidents_of_p_term term
+
+  and filter_bound_qidents
+  (args : Syntax.p_arg list) (terms_list : Syntax.p_term list) =
+    let qids, qargs =  List.split (List.map qidents_of_bound_p_arg args ) in
+    let qids, qargs = List.concat qids, List.concat qargs in
+
+    let args = List.map (fun (id:Syntax.ident) -> id.elt ) qargs in
+
+    let filter_args (id : Syntax.qident) = not (List.mem (snd id.elt) args) in
+
+    let get_qterm term = List.filter filter_args (qidents_of_p_term term) in
+    let qterm = concat_map get_qterm terms_list in
+
+  (*     Format.eprintf "Bound indetifiers :%s\n%!" (String.concat " " args);
+  *)
+    qids@qterm
+
+  and qidents_of_p_config (cfg : Syntax.p_config) = match cfg with
+  | Syntax.P_config_builtin (_, qid) -> [qid]
+  | Syntax.P_config_unop u -> let (_, _, qid) = u in [qid]
+  | Syntax.P_config_binop b -> let (_, _, _, qid) = b in [qid]
+  | Syntax.P_config_ident _ -> []
+  | Syntax.P_config_quant qid -> [qid]
+  | Syntax.P_config_unif_rule rule -> qidents_of_p_rule rule
+
+  and qidents_of_p_rule (rule : Syntax.p_rule) =
+    let (patt, term) = rule.elt in
+    (qidents_of_p_term patt)@(qidents_of_p_term term)
+
+  let qidents_of_cmd (cmd:t) =
+    match cmd.elt with
+    | Syntax.P_require (_,_) -> []
+    | Syntax.P_require_as (_,_) -> []
+    | Syntax.P_open _ -> []
+    | Syntax.P_symbol (_, _, args, term) -> filter_bound_qidents args [term]
+    | Syntax.P_rules rules -> concat_map qidents_of_p_rule rules
+    | Syntax.P_definition (_pl,_b , _location , args ,Some ty, body) ->
+        filter_bound_qidents args [ty; body]
+    | Syntax.P_definition (_pl,_b , _location , args ,None, body) ->
+        filter_bound_qidents args [body]
+    | Syntax.P_theorem (_, statement, _, _) ->
+        let (_, args, body) = statement.elt in
+        filter_bound_qidents args [body]
+    | Syntax.P_set set -> qidents_of_p_config set
+    | Syntax.P_query q -> let f (q : Syntax.p_query_aux) = match q with
+      | Syntax.P_query_verbose _ -> []
+      | Syntax.P_query_debug (_, _) -> []
+      | Syntax.P_query_flag (_, _) -> []
+      | Syntax.P_query_assert (_, assertion) ->
+        let g (assertion : Syntax.p_assertion) = match assertion with
+          | Syntax.P_assert_typing (t1, t2) ->
+              (qidents_of_p_term t1) @ (qidents_of_p_term t2)
+          | Syntax.P_assert_conv (t1, t2) ->
+              (qidents_of_p_term t1) @ (qidents_of_p_term t2)
+        in g assertion
+      | Syntax.P_query_infer (term, _) -> qidents_of_p_term term
+      | Syntax.P_query_normalize (term, _) -> qidents_of_p_term term
+      | Syntax.P_query_prover _ -> []
+      | Syntax.P_query_prover_timeout _ -> []
+
+      in f q.elt
+
+  (*This function is used to get the symbol identifiers
+  in the document and then create a map from them*)
+  let get_qidents (doc : t) = (qidents_of_cmd doc)
+
```

Comment:
Eventually these extraction functions should be moved to their own file, I dunno how to structure the code tho.

Reply:
Done.